### PR TITLE
Allowing pyOpenSSL import to fail for GAE.

### LINKED ISSUE #1445: gcloud-python can't import OpenSSL in App Engine (since it can't be vendored in)
I noticed the depending crypto module was changed since commit 948cccf3. The project was deployed with GAE release 1.9.32 and gcloud-python 0.9.0.

Here are some context info according to your recent comments:

1. How do you vendor in gcloud-python for App Engine?
As described in GAE documentation: https://cloud.google.com/appengine/docs/python/tools/libraries27#vendoring

2. Which services / gcloud-python subpackages do you use? (I'm curious which imports / import paths your code has been using.)
I am generating signed url for an object in Google Cloud Storage. Here's the partial stacktrace:
> Traceback (most recent call last):
> ---- omitted for simple trace) ----
>     from ext import cloudstorage as storage
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/ext/cloudstorage.py", line 8, in <module>
>     from gcloud import (
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/gcloud/storage/__init__.py", line 43, in <module>
>     from gcloud.storage.blob import Blob
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/gcloud/storage/blob.py", line 28, in <module>
>     from gcloud.credentials import generate_signed_url
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/gcloud/credentials.py", line 22, in <module>
>     from OpenSSL import crypto
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/OpenSSL/__init__.py", line 8, in <module>
>     from OpenSSL import rand, crypto, SSL
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/OpenSSL/rand.py", line 11, in <module>
>     from OpenSSL._util import (
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/OpenSSL/_util.py", line 6, in <module>
>     from cryptography.hazmat.bindings.openssl.binding import Binding
>   File "/base/data/home/apps/s~jackend-0/1.390622100755930629/lib/cryptography/hazmat/bindings/openssl/binding.py", line 14, in <module>
>     from cryptography.hazmat.bindings._openssl import ffi, lib
> ImportError: dynamic module does not define init function (init_openssl)

I guess this is caused by the [inherent inflexibility](https://cloud.google.com/appengine/docs/python/tools/libraries27) with runtime environment in GAE, and affects #1009 which seems to conflict with this issue.

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #1445.

### CODE REVIEW

**Comment by jonparrott** [comment]
@dhermes ping me when you're ready for review.

**Comment by dhermes** [comment]
I am "ready" for review. The AppVeyor fail is noted in #1434 

IMO, the coverage drop just needs a pragma no cover. WDYT?

**Comment by jonparrott** [comment]
I think it'd be nice to see a test case for this, if possible.

**Comment by dhermes** [comment]
@jonparrott I have [previously][1] implemented tests that mock import failures and the pay-off does not justify the amount that goes in.

Maybe there is a better way and I just don't know how to do it?

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/0c58512ccafb6aaea8af09c69fa85e955e85084e/gcloud_bigtable/test___init__.py#L19

**Comment by jonparrott** [comment]
I guess I more want to ensure that `_get_pem_key` and `_get_signature_bytes` raises a more useful error than an `AttributeError` if crypto is not available.

**Comment by dhermes** [comment]
OK cool I can add that feature and test for it.

**Comment by dhermes** [comment]
@jonparrott PTAL. (Also remind me to squash to commits before merging.)

File: gcloud/credentials.py [hunk 1]
```diff
@@ -181,7 +187,11 @@ def _get_pem_key(credentials):
         raise TypeError((credentials,
                          'not a valid service account credentials type'))
 
-    return crypto.load_privatekey(crypto.FILETYPE_PEM, pem_text)
+    if crypto is None:
+        raise EnvironmentError('pyOpenSSL must be installed to load a '
```
Comment (jonparrott) [inline gcloud/credentials.py]: nit: I prefer this

```
if crypto is None:
   raise EnvironmentError(
       'pyOpenSSL must be installed to load a private key')
```

But I don't feel strongly enough to make anyone else do it.

File: gcloud/credentials.py [hunk 2]
```diff
@@ -181,7 +187,11 @@ def _get_pem_key(credentials):
...
+                               'private key')
+    else:
```
Comment (jonparrott) [inline gcloud/credentials.py]: Is the else necessary?

File: gcloud/credentials.py [hunk 3]
```diff
@@ -207,7 +218,11 @@ def _get_signature_bytes(credentials, string_to_sign):
         pkey = _get_pem_key(credentials)
         if not isinstance(string_to_sign, six.binary_type):
             string_to_sign = string_to_sign.encode('utf-8')
-        return crypto.sign(pkey, string_to_sign, 'SHA256')
+        if crypto is None:
```
Comment (jonparrott) [inline gcloud/credentials.py]: This check might not actually be necessary as `_get_pem_key` does the same thing and is called at 218. Though, I like the consistency. Your call.

**Comment by jonparrott** [comment]
LGTM with minor nits. You can squash commits if you want, but I don't see an issue with these being separate commits.

File: gcloud/credentials.py (hunk 2 above)
Comment (dhermes) [inline gcloud/credentials.py]: Nope. Will update and ping you.

File: gcloud/credentials.py (hunk 1 above)
Comment (dhermes) [inline gcloud/credentials.py]: Works for me

File: gcloud/credentials.py (hunk 3 above)
Comment (dhermes) [inline gcloud/credentials.py]: Yes I considered that this error would never get a chance to occur in "real life" but put it in anyhow.

These just make me want to upstream this code [into `oauth2client`][1] even more.

[1]: https://github.com/google/oauth2client/blob/b13ab40a01adacf0075fd52f95173abb33297254/oauth2client/service_account.py#L265-L266

Comment (jonparrott) [inline gcloud/credentials.py]: Seems fair.

**Comment by dhermes** [comment]
@jonparrott PTAL

File: gcloud/credentials.py [hunk 4]
```diff
@@ -207,6 +217,9 @@ def _get_signature_bytes(credentials, string_to_sign):
         pkey = _get_pem_key(credentials)
         if not isinstance(string_to_sign, six.binary_type):
             string_to_sign = string_to_sign.encode('utf-8')
+        if crypto is None:
+            raise EnvironmentError('pyOpenSSL must be installed to sign '
```
Comment (jonparrott) [inline gcloud/credentials.py]: same weird formatting here, let's be consistent.

**Comment by jonparrott** [comment]
Good to go with the tiniest little nit. Feel free to merge once that's fixed and travis is happy.

File: gcloud/credentials.py (hunk 4 above)
Comment (dhermes) [inline gcloud/credentials.py]: Weird how? The content doesn't fit on a single line. To be clear, you want me to go from

```python
            raise EnvironmentError('pyOpenSSL must be installed to sign '
                                   'content using a private key')
```

to

```python
            raise EnvironmentError(
                'pyOpenSSL must be installed to sign content using a '
                'private key')
```

Comment (jonparrott) [inline gcloud/credentials.py]: Right, in general I dislike hanging indents. It's up to you- let's just be consistent within the file on whether we use them or not.

Comment (dhermes) [inline gcloud/credentials.py]: OK, will do. I'm going to squash and fold this change into the final (squashed commit) and then will merge once tests pass. LMK if this is no bueno.

Comment (jonparrott) [inline gcloud/credentials.py]: SGTM

**Pull request merged**